executor/predictor/metadata.go

```diff
@@ -0,0 +1,82 @@
+package predictor
+
+import (
+	"github.com/seldonio/seldon-core/operator/apis/machinelearning.seldon.io/v1"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+type MetadataTensor struct {
+	DataType string `json:"datatype,omitempty"`
+	Name     string `json:"name,omitempty"`
+	Shape    []int  `json:"shape,omitempty"`
+}
+
+type ModelMetadata struct {
+	Name     string           `json:"name,omitempty"`
+	Platform string           `json:"platform,omitempty"`
+	Versions []string         `json:"versions,omitempty"`
+	Inputs   []MetadataTensor `json:"inputs,omitempty"`
+	Outputs  []MetadataTensor `json:"outputs,omitempty"`
+}
+
+type GraphMetadata struct {
+	Name         string                   `json:"name"`
+	Models       map[string]ModelMetadata `json:"models"`
+	GraphInputs  []MetadataTensor         `json:"graphinputs"`
+	GraphOutputs []MetadataTensor         `json:"graphoutputs"`
+}
+
+func NewGraphMetadata(p *PredictorProcess, spec *v1.PredictorSpec) (*GraphMetadata, error) {
+	models, err := p.MetadataMap(spec.Graph)
+	if err != nil {
+		return nil, err
+	}
+	output := &GraphMetadata{
+		Name: spec.Name,
+		Models: models,
+	}
+	output.GraphInputs, output.GraphOutputs = output.GetShapeFromGraph(spec.Graph)
+	return output, nil
+}
+
+func (gm *GraphMetadata) GetShapeFromGraph(node *v1.PredictiveUnit) (
+	input []MetadataTensor, output []MetadataTensor,
+) {
+	nodeMeta := gm.Models[node.Name]
+	nodeInputs := nodeMeta.Inputs
+	nodeOutputs := nodeMeta.Outputs
+
+	// Single node graphs: code path terminates here if this is the case
+	if node.Children == nil || len(node.Children) == 0 {
+		// We treat node's inputs/outputs as global despite its Type
+		return nodeInputs, nodeOutputs
+	}
+
+	// Multi nodes graphs
+	if *node.Type == v1.MODEL || *node.Type == v1.TRANSFORMER {
+		// Ignore all children except first one for Models and Transformers
+		_, childOutputs := gm.GetShapeFromGraph(&node.Children[0])
+		return nodeInputs, childOutputs
+	} else if *node.Type == v1.OUTPUT_TRANSFORMER {
+		// Ignore all children except first one for Output Transformers
+		// OUTPUT_TRANSFORMER first passes its input to (first) child and returns the output.
+		childInputs, _ := gm.GetShapeFromGraph(&node.Children[0])
+		return childInputs, nodeOutputs
+	} else if *node.Type == v1.COMBINER {
+		// Combiner will pass request to all of its children and combine their output.
+		// We assume that all children take same type of inputs.
+		childInputs, _ := gm.GetShapeFromGraph(&node.Children[0])
+
+		return childInputs, nodeOutputs
+	} else if *node.Type == v1.ROUTER {
+		// ROUTER will pass request to one of its children and return child's output.
+		// We assume that all children take same type of inputs.
```

Comment:
I see, it's good to state this assumption. I think there are certianly cases where you may want to have different outputs based on the route, but it could be something we can explore further based in how the community uses the graphs. Perhaps one to try to grab some feedback from the community, and potentially tackle as a separate PR. THis assumption should also be made explicitly in the docs

Reply:
Hmm.... Thinking about it again now... In kfserving spec the `inputs` and `outputs` are defined as sequence of the `MetadataTensor`. Our models (as well as transformers) now works by accepting and returning a single tensor, for example
```
      "inputs": [{"datatype": "BYTES", "name": "input", "shape": [1, 5]}],
      "outputs": [{"datatype": "BYTES", "name": "output", "shape": [1, 3]}]
```
I already made use of the fact that inputs/outputs are lists by suggesting to have combiner's input set as 
```
        "inputs": [
            {"name": "input-1", "datatype": "BYTES", "shape": [1, 10]},
            {"name": "input-2", "datatype": "BYTES", "shape": [1, 20]},
        ],
        "outputs": [{"name": "combined output", "datatype": "BYTES", "shape": [3]}],
```

where here it means that it takes input from two children nodes (one sending [1, 10] other [1, 20] shaped output).


What about if router define its meta as:
```
        "inputs": [
            {"name": "input", "datatype": "BYTES", "shape": [1, 10]},
        ],
        "outputs": [
            {"name": "route-1", "datatype": "BYTES", "shape": [1, 3]},
            {"name": "route-2", "datatype": "BYTES", "shape": [1, 5]}
        ]
```

and reports those? It will be up to the system reading that metadata to interpret that in this situation one is not getting both outputs but only one depending on the route taken.

Reply:
It could work even as follows:
- if inputs/outputs not defined on the router node - I derive it from children meta including all output routes as above
- if user define meta on router node - I use that and ignore children's meta

@axsaucedo what do you think?

Reply:
Right... it sounds like there's a couple of things to unpack here. 

First in regards to the current issue of kfserving allowing for multiple tensors, that is certainly an issue, as if we have two combiner children returning multiple tensors each, this would make it ambiguous of which combiner that came from. 

This will also be partially linked to the issue of how we'll deal with the SeldonMessage format, as I know you cannot really have an array of multiple SeldonMessages... 

We may not be able to fully drop the plain tensor structure as we do currently support the tensorflow protocol, so we may have the option to either pass an array of tensors OR a SeldonMessage. 

If we agree that this is the case, then we can then just move into the question of how we can deal with the multiple outputs.

Now in regards to how we can deal with multiple outputs, we will probably ahve to move to a graph level metadata of the following format:

```python
inputs: List[ Union[ List[ Tensor ], SeldonMessage] ],
outputs: List[ Union[ List[ Tensor ], SeldonMessage] ]
```

Which means that for inputs and ouputs we'll have a list of multiple SeldonMessages and/or tensor lists